### PR TITLE
Identifier Triplets RFC-001

### PR DESCRIPTION
**Pull request by ericbuckley**
## Description
Proposing RFC-001, identifier triplets

## Additional Notes
- added a new rfc template for future proposals

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

**Review by cbrinson-rise8 — APPROVED** [review]
Nice! This makes sense to me after reading through it. Great examples too!

**Review by alhayward — COMMENTED** [review]
Love the RFC template! This will be really helpful moving forward as we think critically about and document our design choices. I can create another for our short-term and long-term approaches to possible match, if it's helpful to pull that out of GH Issues. Appreciate the thorough writeup on triplet identifiers, left some questions. Thanks!

File: rfc/000-template.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# RFC: [Title of Proposal]
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply rfc/000-template.md:1]: Request For Change

File: rfc/001-identifier-triplets.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+The type attribute will be limited to a codes defined by the
+[HL7 identifierType code system](https://terminology.hl7.org/6.0.2/CodeSystem-v2-0203.html). This 
+includes roughly 100 different types of identifiers that are all coded using a 2-7 character value.
+For example, `DL` is the code for Driver's License, `SS` is the code for Social Security Number, `MR`
+is the code for Medical Record Number.
+
+The authority attribute will be a free-form string that can be used to specify the issuing authority
+of the identifier.  For example, `CA` could be used to specify that the Driver's License was issued
+by the state of California.  There is no standard code system for this attribute, so it will be up to
+the user to specify a value that makes sense for their data.
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply rfc/001-identifier-triplets.md:106]: If using the `/link` API yes, if using the `/fhir` API then it will be in the identifying value in the [assigner](https://www.hl7.org/fhir/datatypes-definitions.html#Identifier.assigner).

File: rfc/001-identifier-triplets.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+specified type of identifier between the documents is a match.
+
+The evaluation functions will be comparing all 3 parts of the identifier tuple (type, authority,
+value) when determining if two identifiers are a match.  The difference is just between what types
+of identifiers are we going to compare.
+
+```python
+
+if feature == 'IDENTIFIER':
+    assert values == ['DL:CA:A123456', 'SS::123-45-6789']
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply rfc/001-identifier-triplets.md:180]: Yes, if using the `IDENTIFIER` feature matcher.  For that reason, a fuzzy matching function evaluator would likely be a poor choice for comparing different identifiers.  You'd rather do something like exact.  However, if the customer uses a feature matcher of `IDENTIFIER:SS` only the social security numbers will be compared.

File: rfc/001-identifier-triplets.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+require an authority, so we can make a slight reduction in the blocking value size knowing that is
+never required.  In the case of Driver's License, we know that the authority will always be a state,
+so we could implement custom normalization logic attempting to standardize the state values (eg `CA`
+vs `California`).  This is more work long-term, and likely more confusing for users as each identifier
+field has slightly different behavior, but does allow for maximum flexibility.
+
+## Risks and Drawbacks
+
+- Variations in authority values could lead to false negatives in comparisons (eg `CA` vs `California`).
+- Variations in value formats could lead to false negatives in comparisons (eg `123-45-6789` vs `123456789`).
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply rfc/001-identifier-triplets.md:200]: That's a good question, I think we could safely strip leading and trailing whitespace.   I think we might be able to get away with stripping all whitespace characters from the value as well.  I don't think we can safely remove any other values though, as they might have significance to particular types.  There are just too many specified here for me to know the implications it would have on all of them https://terminology.hl7.org/6.0.2/CodeSystem-v2-0203.html.

[the remark this answers is not in the document]
Comment (alhayward) [reply rfc/001-identifier-triplets.md:200]: Sounds good! Sounds like stripping whitespace is sufficient.

**Review by alhayward — COMMENTED** [review]
During our review in the Record Linker <> DIBBs sync on 11/20, here are the questions/feedback Marcelle and Brandon brought up:
* Because the `"authority"` and `"type"` fields are optional, they recommended investigating how missingness in these fields impacts the performance of this approach. For example, if `"authority"` and `"type"` fields are missing a high percentage of the time, how does this concatenation approach perform in terms of match accuracy? Does this approach handle cases of data missingness with sufficient performance?
* Does this approach apply to MRN as well? Or do we have a separate feature matching function for MRN, given it is an important feature with high predictive power?
* If `IDENTIFIER` and `IDENTIFIER:TYPE` will become newly supported features for matching, what will calculating their log odds look like? E.g., will there be one general log odds score for `IDENTIFIER` or one for each `IDENTIFIER:TYPE`?
* For `"authority"` value, Marcelle recommended looking into `Identifier[0].system` in addition to what you document in the RFC, which looks like `Identifier[0].assigner.display`. This should help circumvent the lack of standardization in cases like `"California"` vs. `"CA"` vs. `"DMV of California"`.
* Be sure to use FHIR r4, not r5 (no EHRs currently use r5; one of the links in the RFC redirects to r5 docs)

**Review by cbrinson-rise8 — APPROVED** [review]

**Comment by ericbuckley** [comment]
> During our review in the Record Linker <> DIBBs sync on 11/20, here are the questions/feedback Marcelle and Brandon brought up:
> 
>     * Because the `"authority"` and `"type"` fields are optional, they recommended investigating how missingness in these fields impacts the performance of this approach. For example, if `"authority"` and `"type"` fields are missing a high percentage of the time, how does this concatenation approach perform in terms of match accuracy? Does this approach handle cases of data missingness with sufficient performance?

The `type` and `value` fields will be required, only `authority` is optional.  I'll double-check the RFC to make sure that is clear.

> 
>     * Does this approach apply to MRN as well? Or do we have a separate feature matching function for MRN, given it is an important feature with high predictive power?

The MRN, SSN and DRIVERS_LICENSE features would all be removed and replaced with IDENTIFIER:MR, IDENTIFIER:SS and IDENTIFIER:DL respectively. 
> 
>     * If `IDENTIFIER` and `IDENTIFIER:TYPE` will become newly supported features for matching, what will calculating their log odds look like? E.g., will there be one general log odds score for `IDENTIFIER` or one for each `IDENTIFIER:TYPE`?

Yeah this is an interesting point, I think a log odds score for `IDENTIFIER` might be difficult or impossible to create, so maybe that shouldn't be supported in the log odds type functions.  Only `IDENTIFIER:X` should work.

> 
>     * For `"authority"` value, Marcelle recommended looking into `Identifier[0].system` in addition to what you document in the RFC, which looks like `Identifier[0].assigner.display`. This should help circumvent the lack of standardization in cases like `"California"` vs. `"CA"` vs. `"DMV of California"`.

👍 
> 
>     * Be sure to use FHIR r4, not r5 (no EHRs currently use r5; one of the links in the RFC redirects to r5 docs)

👍

**Review by cbrinson-rise8 — APPROVED** [review]
Looks good and makes sense to me.
